cli/fossilize_replay.cpp

```diff
@@ -2796,20 +2807,38 @@ struct ThreadedReplayer : StateCreatorInterface
 			                 }});
 
 			work.push_back({ get_order_index(ENQUEUE_OUT_OF_RANGE_PARENT_PIPELINES),
-			                 [this, &pipelines, derived]() {
+			                 [this, &pipelines, &parents, derived, parsed_parents]() {
 				                 // Figure out which of the parent pipelines we need.
 				                 for (auto &d : *derived)
-					                 enqueue_parent_pipelines(d, pipelines);
+					                 enqueue_parent_pipelines(d, pipelines, parents, *parsed_parents);
 			                 }});
 
 			if (memory_index == 0)
 			{
 				// This is a join-like operation. We need to wait for all parent pipelines and all shader modules to have completed.
 				work.push_back({get_order_index(ENQUEUE_SHADER_MODULE_SECONDARY_OFFSET),
-				                [this, &parents, hash_offset, start_index]()
+				                [this, &pipelines, &parents, parsed_parents, hash_offset, start_index]()
 				                {
-					                // Wait until all parent pipelines have been parsed.
-					                sync_worker_memory_context(PARENT_PIPELINE_MEMORY_CONTEXT);
+					                unordered_map<Hash, DerivedInfo> top_level_parents = parents;
+
+					                unordered_set<Hash> dependencies;
+					                while (true)
+					                {
+						                // Wait until all parent pipelines have been parsed.
+						                sync_worker_memory_context(PARENT_PIPELINE_MEMORY_CONTEXT);
+
+						                dependencies.clear();
+						                std::swap(*parsed_parents, dependencies);
```

Comment:
It seems flawed to allocate parsed_parents on the outside. We should be clearing parsed parents every iteration, since:

- parents holds the total accumulation of all parsed pipelines
- parsed_parents is an intermediate buffer which holds the pipelines we pushed this iteration

After an iteration is done, parents effectively consumed the information in parsed_parents, so we can safely clear it. It also avoids exploding parse time since we will be parsing the same pipeline for new parents over and over and over. By moving parsed_parents into this lambda and clearing it every iteration, I think we avoid that wart.

Reply:
parsed_parents should already be cleared every iteration by this line:

```
dependencies.clear();
std::swap(*parsed_parents, dependencies);
```

Keeping parse time linear was the reason I introduced parsed_parents in the first place. I allocate parsed_parents outside the lambda, because the first iteration is done in a prior pass.

Reply:
Right